controller/volume_controller.go

```diff
@@ -1553,7 +1652,286 @@ func getRebuildingReplicaCount(e *longhorn.Engine) int {
+		return 0
+	}
+
+	var adjustCount int
+	adjustCount, _, err = fnCount(v, e, rs)
+	if err != nil {
+		return 0
+	}
+	log.Debugf("Found %v replica candidate for auto-balance", adjustCount)
+	return adjustCount
+}
+
+func (vc *VolumeController) getReplicaCountForAutoBalanceBestEffort(v *longhorn.Volume, e *longhorn.Engine,
+	rs map[string]*longhorn.Replica,
+	fnCount replicaAutoBalanceCount) (int, []string, string) {
+	log := getLoggerForVolume(vc.logger, v).WithField("replicaAutoBalanceOption", types.ReplicaAutoBalanceBestEffort)
+
+	var err error
+	defer func() {
+		if err != nil {
+			log.WithError(err).Warn("Skip replica auto-balance")
+		}
+	}()
+
+	setting, err := vc.getAutoBalancedReplicasSetting(v)
+	if err != nil || setting != types.ReplicaAutoBalanceBestEffort {
+		return 0, nil, ""
+	}
+
+	if v.Status.Robustness != types.VolumeRobustnessHealthy {
+		log.Debugf("Cannot auto-balance volume in %s state", v.Status.Robustness)
+		return 0, nil, ""
+	}
+
+	var unusedCount int
+	unusedCount, extraRNames, err := fnCount(v, e, rs)
+	if unusedCount == 0 && len(extraRNames) <= 1 {
+		return 0, nil, ""
+	}
+
+	var mostExtraRList []string
+	var mostExtraRCount int
+	var leastExtraROwner string
+	var leastExtraRCount int
+	for owner, rNames := range extraRNames {
+		rNameCount := len(rNames)
+		if leastExtraRCount == 0 || rNameCount < leastExtraRCount {
+			leastExtraRCount = rNameCount
+			leastExtraROwner = owner
+		}
+		if rNameCount > mostExtraRCount {
+			mostExtraRList = rNames
+			mostExtraRCount = rNameCount
+		}
+	}
+
+	adjustCount := mostExtraRCount - leastExtraRCount
+	if mostExtraRCount == 0 ||
+		(leastExtraRCount != 0 && adjustCount <= 1) {
+		return 0, nil, ""
+	}
+
+	log.Debugf("Found %v replicas to balance to %v from %v", adjustCount, leastExtraROwner, mostExtraRList)
+	return adjustCount, mostExtraRList, leastExtraROwner
+}
+
+func (vc *VolumeController) getReplicaCountForAutoBalanceZone(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) (int, map[string][]string, error) {
+	log := getLoggerForVolume(vc.logger, v).WithField("replicaAutoBalanceType", "zone")
+
+	readyNodes, err := vc.ds.ListReadyAndSchedulableNodes()
+	if err != nil {
+		return 0, nil, err
+	}
+
+	var usedZones []string
+	var usedNodes []string
+	zoneExtraRs := make(map[string][]string)
+	// Count the engine node replica first so it doesn't get included in the
+	// duplicates list.
+	for _, r := range rs {
+		node, exist := readyNodes[r.Spec.NodeID]
+		if !exist {
+			continue
+		}
+		if r.Spec.NodeID == e.Spec.NodeID {
+			nZone := node.Status.Zone
+			zoneExtraRs[nZone] = []string{}
+			usedZones = append(usedZones, nZone)
+			break
+		}
+	}
+	for _, r := range rs {
+		if r.Status.CurrentState != types.InstanceStateRunning {
+			continue
+		}
+
+		node, exist := readyNodes[r.Spec.NodeID]
+		if !exist {
+			log.Debugf("Replica %v on node %v not count for auto-balance, could get evicted", r.Spec.NodeID, r.Name)
```

Comment:
The unnecessary debug logs here and below.

Reply:
Addressed.